servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java

```diff
@@ -0,0 +1,547 @@
+        this.maxMappedDemand = maxMappedDemand;
+        this.maxConcurrency = maxConcurrency;
+        this.delayError = delayError;
+    }
+
+    @Override
+    public Subscriber<? super T> apply(final Subscriber<? super R> subscriber) {
+        return new FlatMapSubscriber<>(this, subscriber);
+    }
+
+    private static final class FlatMapSubscriber<T, R> implements Subscriber<T>, Subscription {
+        private static final Object MAPPED_SOURCE_COMPLETE = new Object();
+        @SuppressWarnings("rawtypes")
+        private static final AtomicReferenceFieldUpdater<FlatMapSubscriber, CompositeException> delayedErrorUpdater =
+                newUpdater(FlatMapSubscriber.class, CompositeException.class, "delayedError");
+        @SuppressWarnings("rawtypes")
+        private static final AtomicIntegerFieldUpdater<FlatMapSubscriber> emittingLockUpdater =
+                AtomicIntegerFieldUpdater.newUpdater(FlatMapSubscriber.class, "emittingLock");
+        @SuppressWarnings("rawtypes")
+        private static final AtomicLongFieldUpdater<FlatMapSubscriber> pendingDemandUpdater =
+                AtomicLongFieldUpdater.newUpdater(FlatMapSubscriber.class, "pendingDemand");
+        @SuppressWarnings("rawtypes")
+        private static final AtomicIntegerFieldUpdater<FlatMapSubscriber> activeMappedSourcesUpdater =
+                AtomicIntegerFieldUpdater.newUpdater(FlatMapSubscriber.class, "activeMappedSources");
+        @SuppressWarnings("rawtypes")
+        private static final AtomicReferenceFieldUpdater<FlatMapSubscriber, TerminalNotification>
+                terminalNotificationUpdater = newUpdater(FlatMapSubscriber.class, TerminalNotification.class,
+                "terminalNotification");
+
+        @SuppressWarnings("unused")
+        @Nullable
+        private volatile TerminalNotification terminalNotification;
+        @Nullable
+        private volatile CompositeException delayedError;
+        @SuppressWarnings("unused")
+        private volatile int emittingLock;
+        private volatile int activeMappedSources;
+        private volatile long pendingDemand;
+
+        // protected by emitting lock, or only accessed in side the Subscriber thread
+        private boolean targetTerminated;
+        private int upstreamDemand;
+        @Nullable
+        private Subscription subscription;
+
+        private final Subscriber<? super R> target;
+        private final Queue<Object> signals;
+        private final PublisherFlatMapMerge<T, R> source;
+        private final Set<FlatMapPublisherSubscriber<T, R>> subscribers;
+
+        FlatMapSubscriber(PublisherFlatMapMerge<T, R> source, Subscriber<? super R> target) {
+            this.source = source;
+            this.target = target;
+            // Start with a small capacity as maxConcurrency can be large.
+            signals = newUnboundedMpscQueue(min(2, source.maxConcurrency * source.maxMappedDemand));
+            subscribers = newSetFromMap(new ConcurrentHashMap<>());
```

Comment:
nit: should we use some "non default" initial capacity ?

Reply:
sure will put a cap on the initial capacity.